linkifiers: Joint Validation

aero31aero opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Addresses #12323.

**Testing Plan:** <!-- How have you tested? -->
Modifies automated tests, manual testing.

aero31aero commented:
This is ready for review. It appears that we still have the casper flake in admin.js

timabbott commented on zerver/models.py:
```diff
@@ -698,6 +698,41 @@ class RealmFilter(models.Model):
     class Meta:
         unique_together = ("realm", "pattern")
 
+    def clean(self) -> None:
```
Can you update this commit message to explain to what extent this commit's adding of a `clean` method might reduce how much validation is done?  I don't remember enough details on how Django's model system `clean` methods work to be sure if this is a strict addition or not at this stage.

aero31aero replied:
This is a strict addition. We haven't removed any previous validations yet; they are removed in the next commit (`realm_filters: Remove regex based method for url and patern validation.`).

Django's full_clean that we call before saving a realm_filter in actions.py [does](https://kite.com/python/docs/django.db.models.Model.full_clean) the following:

> This method calls Model.clean_fields(), Model.clean(), and Model.validate_unique() (if validate_unique is True), in that order and raises a ValidationError that has a message_dict attribute containing errors from all three stages.

timabbott commented on zerver/models.py:
```diff
@@ -698,6 +698,41 @@ class RealmFilter(models.Model):
+        if len(group_list) > 1:
+            group_list.sort()
+            raise ValidationError(_("Groups '{}' in filter pattern is not present "
+                                    "in URL format string.".format(', '.join(group_list))))
```
I think it's an unnecessary cost to translators to have two variants here; it's extra work for translators.  I don't think it's critical to list all groups; so I'd prefer to refactor these blocks to deduplicate the strings for translator convenience (and just raise about the first item in `group_list`.

aero31aero replied:
Fixed.

timabbott commented on zerver/models.py:
```diff
@@ -698,6 +698,41 @@ class RealmFilter(models.Model):
+        try:
+            URLValidator()(substituted)
+        except ValidationError:
+            raise ValidationError(_('Filter pattern does not result in valid URL.'))
```
Can you add a test covering this corner case?  OR is it not reachable at this commit's stage because of the other validation we do?

aero31aero replied:
It's not reachable yet. I added the testcase for this in the commit that removes the other validation and makes this code reachable.

timabbott commented on zerver/tests/test_realm_filters.py:
```diff
@@ -27,7 +27,7 @@ def test_create(self) -> None:
         data = {"pattern": "", "url_format_string": "https://realm.com/my_realm_filter/%(id)s"}
         result = self.client_post("/json/realm/filters", info=data)
-        self.assert_json_error(result, 'This field cannot be blank.')
+        self.assert_all_json_errors(result, ['This field cannot be blank.', 'No groups found in URL format string.'])
```
Does the frontend handle multiple errors?

aero31aero replied:
Yes.

---

![image](https://user-images.githubusercontent.com/8033238/73116989-3adc8a00-3f65-11ea-993d-c05968794382.png)


This is the result with the API returning:

```json
{
    "result":"error",
    "msg":"This field cannot be blank.",
    "errors":{
        "pattern":[
            "This field cannot be blank."
        ],
        "url_format_string":[
            "This field cannot be blank."
        ],
        "__all__":[
            "No groups found in URL format string."
        ]
    }
}
```

timabbott commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -335,8 +335,8 @@
-      "input": "link: xx, x xxxxx xx xxxx xx\n\n[xxxxx #xx](http://xxxxxxxxx:xxxx/xxx/xxxxxx%xxxxxx/xx/):**xxxxxxx**\n\nxxxxxxx xxxxx xxxx xxxxx:\n`xxxxxx`: xxxxxxx\n`xxxxxx`: xxxxx\n`xxxxxx`: xxxxx xxxxx",
-      "expected_output": "<p>link: xx, x xxxxx xx xxxx xx</p>\n<p><a href=\"http://xxxxxxxxx:xxxx/xxx/xxxxxx%xxxxxx/xx/\" target=\"_blank\" title=\"http://xxxxxxxxx:xxxx/xxx/xxxxxx%xxxxxx/xx/\">xxxxx #xx</a>:<strong>xxxxxxx</strong></p>\n<p>xxxxxxx xxxxx xxxx xxxxx:<br>\n<code>xxxxxx</code>: xxxxxxx<br>\n<code>xxxxxx</code>: xxxxx<br>\n<code>xxxxxx</code>: xxxxx xxxxx</p>",
+      "input": "link: xx, x xxxxx xx xxxx xx\n\n[xxxxx #xx](http://xxxxxxxxx.com:1234/xxx/xxxxxx%xxxxxx/xx/):**xxxxxxx**\n\nxxxxxxx xxxxx xxxx xxxxx:\n`xxxxxx`: xxxxxxx\n`xxxxxx`: xxxxx\n`xxxxxx`: xxxxx xxxxx",
+      "expected_output": "<p>link: xx, x xxxxx xx xxxx xx</p>\n<p><a href=\"http://xxxxxxxxx.com:1234/xxx/xxxxxx%xxxxxx/xx/\" target=\"_blank\" title=\"http://xxxxxxxxx.com:1234/xxx/xxxxxx%xxxxxx/xx/\">xxxxx #xx</a>:<strong>xxxxxxx</strong></p>\n<p>xxxxxxx xxxxx xxxx xxxxx:<br>\n<code>xxxxxx</code>: xxxxxxx<br>\n<code>xxxxxx</code>: xxxxx<br>\n<code>xxxxxx</code>: xxxxx xxxxx</p>",
```
Can you explain why these test cases needed to be updated in the commit message?  I don't understand what these specific tests have to do with the changes being made.

aero31aero replied:
I'll add the explanation in the commit message as well. The additional netloc validation fails with `xxxx` as the port number and `xxxxxxx` as the domain name.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -113,6 +113,32 @@ def normal_compile(pattern: str) -> Any:
+    host_re = '(' + hostname_re + domain_re + tld_re + '|localhost)'
+
+    netloc_re = r'(?:' + ipv4_re + '|' + ipv6_re + '|' + host_re + ')' + r'(?::\d{2,5})?'
+    return verbose_compile(netloc_re)
```
Can you explain in comments in this function what is changed from the Django implementation?  Or is this unmodified, and just extracted because we want to use it?  

In any case, I think we should move this function to a new `zerver/lib/url_validator.py` for clarity.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1433,6 +1459,11 @@ def sanitize_url(url: str) -> Optional[str]:
+    # Check for invalid domain names.
+    netloc_re = get_compiled_netloc_re()
+    if netloc and not netloc_re.match(netloc):
+        return None
```
Can you add some tests covering the invalid domain name cases?

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1576,21 +1607,25 @@ def prepare_realm_pattern(source: str) -> str:
+        substituted_url = self.format_string % m.groupdict()
+        url_element = url_to_a(db_data, substituted_url, m.group(OUTER_CAPTURE_GROUP))
+        if isinstance(url_element, str):
+            return None, None, None
```
This probably deserves a comment explaining why the result being a `str` means that it's a failed match.

timabbott commented:
Overall, this looks like a promising direction.  I posted a bunch of comments above on details; I think the rest of the security argument for this change should be written down explicitly.  I think it's something like:

* On the backend, linkifiers only allow the user to create links to things they could have added an explicit markdown link to anyway.  
* <something about how we run linkifiers in the frontend markdown processor? also being safe with this change?>.  We should analyze `exports.set_realm_filters` with these changes in mind.  One possible conclusion is this will reduce the effectiveness of the Python/JS logic.
* This change allows running a wider range of untrusted regular expressions in this code path.  
* The previous code did not prevent DoS through a very expensive regular expression.  The impact of such a regular expression DoS would be primarily to DoS the organization they are an administrator for (since we have timeout code to prevent it from affecting other organizations), which they could have done more effectively using their permissions by deactivating the organization instead.

aero31aero commented:
@timabbott I've addressed the comments so far. This is tangentially related to #13076 so I'm inclined to update the docs on linkification after both these are merged and we have a decision on whether to support custom schemes in realm filters.

(aside: we should probably agree on one term ammongst realm patterns, realm filters, etc.)

timabbott commented:
For terminology, `linkifiers` is what these should be called.  We have a legacy situation that the codebase calls them `realm_filters`, but my hope is that we'll be able to rename that in the code/API eventually.  It's a bit awkward to do so in the API because clients like mobile are relying on the structure, but we could usefully migrate all code other than the API.

jamesvl commented:
Apart from the rebase to resolve merge conflicts, does this PR need any further changes?

I'm willing to pitch in and help out, since I'd like to see the bug fix, but I'm not sure from the final comments what work remains.

aero31aero commented:
@jamesvl the code is ready but we have some discussions left to do. I expect this will make it into the next release.

jamesvl commented:
Just following up again @aero31aero - looks like the merge conflict is just a deleted file. Let me know if I can help in any way to get this rebased and backed for consideration in a release.

timabbott commented:
@aero31aero can you investigate the Puppeteer frontend test failure?  It's in the settings UI for this feature.

timabbott commented on zerver/models.py:
```diff
@@ -773,6 +773,45 @@ class RealmFilter(models.Model):
+        except Exception:
+            # regex validation errors would be caught by full_clean()
+            # so we can safely just return here.
+            return
```
This feels unsafe; can we just raise a validation error here?

timabbott replied:
Or maybe better, just remove the field-level validator and do the clean check here?

andersk commented on zerver/models.py:
```diff
@@ -773,6 +773,45 @@ class RealmFilter(models.Model):
+        # extract substitutions
+        try:
+            pattern = re.compile(self.pattern)
+        except Exception:
```
Better to be specific (`except re.error`).

andersk commented on zerver/models.py:
```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
+        for m in self.group_match_regex.finditer(self.url_format_string):
+            name = m.group('group_name')
+            if name not in group_list:
+                raise ValidationError(_("Group '{}' in URL format string is "
+                                        "not present in filter pattern.".format(name)))
```
Incorrect use of `_()`; the `.format` needs to be outside. I wonder how you evaded our semgrep check for this?

andersk commented on zerver/models.py:
```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
+            # resolve errors if there are multiple missing patterns.
+            group_list.sort()
+            group_name = group_list[0]
+            raise ValidationError(_("Group '{}' in filter pattern is not present "
+                                    "in URL format string.".format(group_name)))
```
Also incorrect use of `_()`.

andersk commented on zerver/tests/test_markdown.py:
```diff
@@ -1032,16 +1032,23 @@ def test_realm_patterns(self) -> None:
             if should_have_converted:
-                self.assertTrue('https://trac.example.com' in converted)
+                self.assertTrue('https://' in converted)
+                self.assertTrue('trac.example.com' in converted)
```
This obviously isn’t a real vulnerability in the context of tests, but GitHub code scanning [doesn’t like this pattern](https://github.com/github/codeql/blob/main/python/ql/src/Security/CWE-020/IncompleteUrlSubstringSanitization.ql), which seems like a good reason to avoid it if it’s not too hard.

andersk commented on zerver/lib/url_validator.py:
```diff
@@ -0,0 +1,36 @@
+    '''
+    From: https://docs.djangoproject.com/en/3.0/_modules/django/core/validators/#URLValidator
+
+    This method takes code from Django's URLValidator. We cannot use the URLValidator as is
+    because it performs checks on the scheme as well and we have code to deal with schemes
+    already that doesn't agree with URLValidator. Thus, we extract the netloc specific parts
+    of URLValidator here.
```
Do we really have to? The upstream `URLValidator` constructor accepts an optional list of schemes, so I’d think we should be able to use it directly.

andersk commented on zerver/models.py:
```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
+            return
+        group_list = list(pattern.groupindex.keys())
+
+        # Report regexes without any patterns.
+        if len(group_list) == 0:
+            raise ValidationError(_("No groups found in URL format string."))
```
Is there actually a reason to forbid this? I could see use cases for linkifiers that just match a constant string.

andersk commented on zerver/models.py:
```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
+            if name not in group_list:
+                raise ValidationError(_("Group '{}' in URL format string is "
+                                        "not present in filter pattern.".format(name)))
+            group_list.remove(name)
```
`list.remove` is a linear-time operation, so this loop has quadratic complexity. Use a `set`.

andersk commented on zerver/models.py:
```diff
@@ -768,11 +762,55 @@ class RealmFilter(models.Model):
+            group_list.remove(name)
+
+        # Report patterns missing in URL format string.
+        if len(group_list) >= 1:
```
PEP 8 says this should be `if group_list:`.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1616,23 +1627,35 @@ def prepare_realm_pattern(source: str) -> str:
     return fr"""(?<![^\s'"\(,:<])(?P<{OUTER_CAPTURE_GROUP}>{source})(?!\w)"""
 
+# Convert user inputted %()s realm filter strings to be {} type strings
+def convert_to_format_string(percent_string: str) -> str:
+    regex = RealmFilter.group_match_regex
+    return regex.sub(lambda m: "{{{}}}".format(m.group('group_name')), percent_string)
```
This will break on `%`-formatting strings containing `%%`, `{`, or `}`, and calls for more principled parsing than regexes can reasonably provide.

timabbott commented:
As discussed in https://github.com/zulip/zulip/issues/12323, this issue was finally resolved through other PRs. Thanks @aero31aero!
